tools/serverpod_cli/lib/src/util/latest_cli_version.dart

```diff
@@ -0,0 +1,101 @@
+import 'dart:convert';
+
+import 'package:http/http.dart' as http;
+import 'package:pub_semver/pub_semver.dart';
+import 'package:serverpod_cli/src/downloads/resource_manager.dart';
+
+abstract class LatestCliVersionConstants {
+  static const pubDevUri = 'https://pub.dev/api/packages/serverpod_cli';
+  static const badConnectionRetryTimeout = Duration(hours: 1);
+  static const localStorageValidityTime = Duration(days: 1);
+  static const pubDevConnectionTimeout = Duration(seconds: 2);
+}
+
+/// Attempts to fetch the latest cli version from either local storage of pub.dev
+Future<Version?> tryFetchLatestValidCliVersion({
+  localStorageService = const LocalStorageService(),
+  pubDevService = const PubDevService(),
+}) async {
+  var latestCliVersion = await localStorageService.fetchLatestCliVersion();
+
+  if (latestCliVersion != null) return latestCliVersion;
+
+  latestCliVersion = await pubDevService.tryFetchAndStoreLatestVersion();
+
+  return latestCliVersion;
+}
+
+class LocalStorageService {
+  final String? optionalLocalStoragePath;
+
+  const LocalStorageService({this.optionalLocalStoragePath});
+  LocalStorageService.nonConst({this.optionalLocalStoragePath});
+
+  Future<Version?> fetchLatestCliVersion() async {
+    var localStorageData = await resourceManager.tryFetchLatestCliVersion(
+      localStoragePath: optionalLocalStoragePath,
+    );
+    if (localStorageData == null) return null;
+
+    return localStorageData.validUntil.isAfter(DateTime.now())
+        ? localStorageData.version
+        : null;
+  }
+}
+
+class PubDevService {
+  final String? optionalLocalStoragePath;
+  final http.Client? client;
+  final Duration timeout;
+
+  const PubDevService({
+    this.optionalLocalStoragePath,
+    this.client,
+    this.timeout = LatestCliVersionConstants.pubDevConnectionTimeout,
+  });
+  PubDevService.nonConst({
+    this.optionalLocalStoragePath,
+    this.client,
+    this.timeout = LatestCliVersionConstants.pubDevConnectionTimeout,
+  });
+
+  Future<Version?> tryFetchAndStoreLatestVersion() async {
+    var pubDevLatestVersion = await _tryFetchLatestCliVersion();
+
+    LatestCliVersionArtefact versionArtefact;
+    Version? returnVersion;
+    if (pubDevLatestVersion != null) {
+      versionArtefact = LatestCliVersionArtefact(
+        pubDevLatestVersion,
+        DateTime.now().add(LatestCliVersionConstants.localStorageValidityTime),
+      );
+      returnVersion = pubDevLatestVersion;
+    } else {
+      versionArtefact = LatestCliVersionArtefact(
+        Version.none,
+        DateTime.now().add(LatestCliVersionConstants.badConnectionRetryTimeout),
+      );
+    }
+
+    await resourceManager.storeLatestCliVersion(versionArtefact,
+        localStoragePath: optionalLocalStoragePath);
+    return returnVersion;
+  }
+
+  Future<Version?> _tryFetchLatestCliVersion() async {
+    try {
```

Comment:
Have you checked if there is an official dart package for doing this? I feel like there may be some package for interacting with pub.dev. (Could be safer to use Google's own stuff)

Reply:
As far as I could tell, the dart team unfortunately doesn't have an official tool for this. I thought [cli_util](https://pub.dev/packages/cli_util) would help but it doesn't support fetching the version.

These are other packages I found:
- [pub_updater](https://pub.dev/packages/pub_updater)
- [cli_notify](https://pub.dev/packages/cli_notify)
- [update](https://pub.dev/packages/update)
They have unfortunately not reached version 1 yet so I'm not sure if it is worth adding a dependency for such simple logic in that case.

As a reference the flutter team has implemented their [own version of this](https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/version.dart), but then checking towards a GitHub repo instead of pub.dev: